File: backend/store.py
"""
Persistent JSON store for OpenClaw sandbox records.
Writes to data/sandboxes.json by default.
"""

import json
import os
from datetime import datetime, timezone
from pathlib import Path
from typing import Any

STORE_PATH = Path(os.getenv("SANDBOXES_FILE", Path(__file__).parent / "data" / "sandboxes.json"))


def _read() -> dict[str, Any]:
    if not STORE_PATH.exists():
        return {}
    try:
        return json.loads(STORE_PATH.read_text())
    except (json.JSONDecodeError, OSError):
        return {}


def _write(data: dict[str, Any]) -> None:
    STORE_PATH.write_text(json.dumps(data, indent=2))


def save_sandbox(result: dict[str, Any], sandbox_name: str = "") -> None:
    """Upsert a sandbox record keyed by sandbox_id."""
    store = _read()
    sandbox_id = result["sandbox_id"]
    existing = store.get(sandbox_id, {})

    store[sandbox_id] = {
        **existing,
        "sandbox_id": sandbox_id,
        "sandbox_name": sandbox_name or result.get("sandbox_name", "openclaw-gateway"),
        "sandbox_state": result.get("sandbox_state", ""),
        "dashboard_url": result.get("dashboard_url"),
        "signed_url": result.get("signed_url"),
        "standard_url": result.get("standard_url"),
        "preview_token": result.get("preview_token"),
        "gateway_token": result.get("gateway_token"),
        "gateway_port": result.get("gateway_port", 18789),
        "ssh_command": result.get("ssh_command", ""),
        "created_at": existing.get("created_at") or datetime.now(timezone.utc).isoformat(),
        "approved": existing.get("approved", False),
    }
    _write(store)


def mark_approved(sandbox_id: str) -> None:
    store = _read()
    if sandbox_id in store:
        store[sandbox_id]["approved"] = True
        _write(store)


def list_sandboxes() -> list[dict[str, Any]]:
    store = _read()
    return sorted(store.values(), key=lambda s: s.get("created_at", ""), reverse=True)


def delete_sandbox(sandbox_id: str) -> bool:
    store = _read()
    if sandbox_id not in store:
        return False
    del store[sandbox_id]
    _write(store)
    return True


def get_sandbox(sandbox_id: str) -> dict[str, Any] | None:
    return _read().get(sandbox_id)

File: ruh-backend/src/store.ts
/**
 * PostgreSQL-backed store for OpenClaw sandbox records.
 */

import { withConn } from './db';

export interface SandboxRecord {
  sandbox_id: string;
  sandbox_name: string;
  sandbox_state: string;
  dashboard_url: string | null;
  signed_url: string | null;
  standard_url: string | null;
  preview_token: string | null;
  gateway_token: string | null;
  gateway_port: number;
  vnc_port: number | null;
  ssh_command: string;
  created_at: string;
  approved: boolean;
  shared_codex_enabled: boolean;
  shared_codex_model: string | null;
}

export async function saveSandbox(
  result: Record<string, unknown>,
  sandboxName = '',
): Promise<void> {
  await withConn(async (client) => {
    await client.query(
      `
      INSERT INTO sandboxes (
        sandbox_id, sandbox_name, sandbox_state, dashboard_url,
        signed_url, standard_url, preview_token, gateway_token,
        gateway_port, vnc_port, ssh_command, shared_codex_enabled, shared_codex_model
      ) VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10, $11, $12, $13)
      ON CONFLICT (sandbox_id) DO UPDATE SET
        sandbox_name  = EXCLUDED.sandbox_name,
        sandbox_state = EXCLUDED.sandbox_state,
        dashboard_url = EXCLUDED.dashboard_url,
        signed_url    = EXCLUDED.signed_url,
        standard_url  = EXCLUDED.standard_url,
        preview_token = EXCLUDED.preview_token,
        gateway_token = EXCLUDED.gateway_token,
        gateway_port  = EXCLUDED.gateway_port,
        vnc_port      = EXCLUDED.vnc_port,
        ssh_command   = EXCLUDED.ssh_command,
        shared_codex_enabled = EXCLUDED.shared_codex_enabled,
        shared_codex_model = EXCLUDED.shared_codex_model
      `,
      [
        result['sandbox_id'],
        sandboxName || result['sandbox_name'] || 'openclaw-gateway',
        result['sandbox_state'] ?? '',
        result['dashboard_url'] ?? null,
        result['signed_url'] ?? null,
        result['standard_url'] ?? null,
        result['preview_token'] ?? null,
        result['gateway_token'] ?? null,
        result['gateway_port'] ?? 18789,
        typeof result['vnc_port'] === 'number' ? result['vnc_port'] : null,
        result['ssh_command'] ?? '',
        Boolean(result['shared_codex_enabled']),
        typeof result['shared_codex_model'] === 'string' ? result['shared_codex_model'] : null,
      ],
    );
  });
}

export async function markApproved(sandboxId: string): Promise<void> {
  await withConn(async (client) => {
    await client.query(
      'UPDATE sandboxes SET approved = TRUE WHERE sandbox_id = $1',
      [sandboxId],
    );
  });
}

export async function updateSandboxSharedCodex(
  sandboxId: string,
  enabled: boolean,
  model: string | null,
): Promise<void> {
  await withConn(async (client) => {
    await client.query(
      `UPDATE sandboxes
       SET shared_codex_enabled = $2,
           shared_codex_model = $3,
           sandbox_state = COALESCE(NULLIF(sandbox_state, ''), 'running')
       WHERE sandbox_id = $1`,
      [sandboxId, enabled, model],
    );
  });
}

export async function listSandboxes(): Promise<SandboxRecord[]> {
  return withConn(async (client) => {
    const res = await client.query(
      'SELECT * FROM sandboxes ORDER BY created_at DESC',
    );
    return res.rows.map(serialize);
  });
}

export async function getSandbox(sandboxId: string): Promise<SandboxRecord | null> {
  return withConn(async (client) => {
    const res = await client.query(
      'SELECT * FROM sandboxes WHERE sandbox_id = $1',
      [sandboxId],
    );
    return res.rows.length > 0 ? serialize(res.rows[0]) : null;
  });
}

export async function deleteSandbox(sandboxId: string): Promise<boolean> {
  return withConn(async (client) => {
    await client.query(
      'DELETE FROM conversations WHERE sandbox_id = $1',
      [sandboxId],
    );
    const res = await client.query(
      'DELETE FROM sandboxes WHERE sandbox_id = $1',
      [sandboxId],
    );
    return (res.rowCount ?? 0) > 0;
  });
}

function serialize(row: Record<string, unknown>): SandboxRecord {
  if (row['created_at'] instanceof Date) {
    row['created_at'] = row['created_at'].toISOString();
  }
  return row as unknown as SandboxRecord;
}
